Adding graphql

Danila-Peshev opened the pull request:
Adding graphql to backend

Danila-Peshev committed:
installed apollo-server, graphql.

Danila-Peshev committed:
Commit:
A small change in the getMoviesByFilters method of the service;
Added Apollo, the schema is generated automatically based on graphql mutation queries and classes;
Added the Auth and Resolver block for the same;
Graphql classes have been created to type the return values of queries in the tmdb service;
The service's work with the user has been updated;
Added a resolver for the user.

Isidora25 commented on src/auth/auth.resolver.ts:
```diff
@@ -0,0 +1,17 @@
+    @Args('password', { type: () => String }) password: string,
+  ) {
+    const optUser = await this.authService.login(email, password);
+    return optUser;
```
```suggestion
    return this.authService.login(email, password);
    
```

Isidora25 commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,27 @@
+      return null
+    }
+  }
+}
```
```suggestion
  async login(email: string, password: string): Promise<User> {
  const user = await this.userService.findByEmail(email);
    const isPasswordEquals = password === user?.password;
    if (!user ||  !isPasswordEquals) {
      throw new UnauthorizedException();
    } 
    return user
  }
}
```

Isidora25 commented on src/tmdb-movie/movies.resolver.ts:
```diff
@@ -0,0 +1,70 @@
+    @Args('releaseYear', { type: () => Int, nullable: true })
+    releaseYear?: number,
+    @Args('page', { type: () => Int, nullable: true }) page?: number,
+  ): Promise<MoviesResponse> {
```
чтобы не перечислять каждый аргумент, если их много, можно передавать параметры скопом 
https://docs.nestjs.com/graphql/resolvers#dedicated-arguments-class

Isidora25 commented on src/user/user.resolver.ts:
```diff
@@ -0,0 +1,74 @@
+    @Args('userId', { type: () => Int }) userId: number,
+  ) {
+    return (await this.userService.findAllGenresByUserId(userId)) || [];
+  }
```
смысловой разницы между getAllGenres  и getGenres нет, зато getGenres короче

Isidora25 commented on src/user/user.resolver.ts:
```diff
@@ -0,0 +1,74 @@
+    @Args('userId', { type: () => Int }) userId: number,
+  ) {
+    return await this.userService.findAllMoviesByUserId(userId);
+  }
```
я бы посоветовала не хранить в user resolver ничего, что не возвращает юзера или не меняет его 
представь, у тебя появилось 10 новых сущностей в приложении, которые связаны с юзером как-то и для каждой сущности тебе нужно написать CRUD
твой резолвер уже сейчас не маленький, а при масштабировании вообще разрастется до размеров кратера на марсе 
так что держи все что относится к отдельному домену в директории этого домена 
фильмы к фильмам, жанры к жанрам

Isidora25 commented on src/user/user.resolver.ts:
```diff
@@ -0,0 +1,74 @@
+    @Args('userId', { type: () => Int }) userId: number,
+    @Args('movieId', { type: () => Int }) movieId: number,
+  ) {
+    return await this.userService.watchByMovieIdToUserByUserId(userId, movieId);
```
watchByMovieIdToUserByUserId - очень сложное название 
если ты напишешь markMovieWatched или что-то подобное, то смысл не поменяется, но название станет короче и чище 
пометка byMovieId или byUserId применима в том случае, когда у тебя есть альтернативы 
например, byUserEmail или что-то подобное 
у тебя же все просто пока, никаких проблем с однозначной идентификацией не дожно быть 
так что старайся Keep It Simple Stupid 
а когда понадобится, расширишь название 

ну YAGNI тут тоже применим

Isidora25 commented on src/user/user.resolver.ts:
```diff
@@ -0,0 +1,74 @@
+    @Args('movieId', { type: () => Int }) movieId: number,
+  ) {
+    return await this.userService.removeMovieIdToUserByUserId(userId, movieId);
+  }
```
можно объединить и сделать toggleMovie

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+
+  async findByEmail(email: string): Promise<User> {
+    return await this.userRepository.findOneBy({ email });
+  }
```
```suggestion
  async findByEmail(email: string): Promise<User> {
    return this.userRepository.findOneBy({ email });
  }
```

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+      user: { id: userId },
+    });
+    return genres;
+  }
```
```suggestion
  async findAllGenresByUserId(userId: number): Promise<FavoriteGenre[]> {
    return = this.genreRepository.findBy({
      user: { id: userId },
    });
    return genres;
  }
```

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+    const movies: FavoriteMovie[] = await this.movieRepository.findBy({
+      user: { id: userId },
+    });
+    return movies;
```
.

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+    userId: number,
+    genreId: number,
+  ): Promise<FavoriteGenre> {
+    const user: User = await this.userRepository.findOneBy({ id: userId });
```
тебе не нужно делаь type assertion каждый раз 
typeorm довольно умный, чтобы возвращать нужный тип

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+    return movies;
+  }
+
+  async addGenreIdToUserByUserId(
```
снова перегруженное название, проверь все названия сервисов

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+      where: { user: { id: userId }, genreId },
+    });
+    if (!existingFavoriteGenre) {
+      const newFavoriteGenre = this.genreRepository.create({
```
```suggestion
      const newFavoriteGenre = await this.genreRepository.create({
```

Danila-Peshev replied:
Когда добавляю ts предупреждает:
'await' has no effect on the type of this expression.

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+    });
+    if (!existingFavoriteGenre) {
+      const newFavoriteGenre = this.genreRepository.create({
+        user,
```
тебе не нужно кидать весь объект юзера при сохранении, должно быть достаточно userId для создания связи

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+    });
+    if (existingFavoriteGenre) {
+      await this.genreRepository.delete({ genreId });
+      return 'Genre has been removed';
```
лучше создать какой-то общий класс для ответов в случае успеха мутации, что-то типа {success: true}

Isidora25 commented on src/user/user.service.ts:
```diff
@@ -0,0 +1,121 @@
+    });
+    if (existingFavoriteMovie) {
+      existingFavoriteMovie.isWatched = true;
+      return await this.movieRepository.save(existingFavoriteMovie);
```
existingFavoriteMovie.save() тоже должно работать

Danila-Peshev replied:
Нет, ругается что отсутствует сущность

Danila-Peshev committed:
add args types to movies resolver.

Danila-Peshev committed:
Commit:
Combined the viewed and not watched movie functions into one;
Rename functions, queries and mutations.

Danila-Peshev committed:
Interaction with genres is separate

Danila-Peshev committed:
Commit:
Interaction with movies is separate;
Delete user resolver.

Danila-Peshev committed:
Shorted code.

Danila-Peshev committed:
Add mutation result graphql class to return functions whose return value is not provided.

Isidora25 commented on src/movie/favorite-movie.service.ts:
```diff
@@ -0,0 +1,59 @@
+    return {success: false}
+  }
+
+  async toggleUserMovie(
```
а вот тут непонятно 
toggleWatchMovie

Isidora25 commented on src/movie/favorite-movie.service.ts:
```diff
@@ -0,0 +1,59 @@
+    });
+  }
+
+  async addUserMovie(userId: number, movieId: number): Promise<FavoriteMovie> {
+    const existingFavoriteMovie = await this.movieRepository.findOne({
+      where: { user: { id: userId }, movieId },
+    });
+    if (!existingFavoriteMovie) {
+      const newFavoriteMovie = this.movieRepository.create({
+        user: { id: userId },
+        movieId,
+        isWatched: false,
+      });
+      return await this.movieRepository.save(newFavoriteMovie);
+    }
+  }
+
+  async removeUserMovie(userId: number, movieId: number): Promise<MutationResult> {
+    const existingFavoriteMovie = await this.movieRepository.findOne({
+      where: { user: { id: userId }, movieId },
+    });
+    if (existingFavoriteMovie) {
+      await this.movieRepository.delete({ movieId });
+      return {success: true};
+    }
+    return {success: false}
+  }
```
это тоже можно объединить

Danila-Peshev committed:
Rename watch unwatch function and query

Danila-Peshev committed:
Commit:
Corrected the switching of watching a movie in the service;
Combined addition and deletion in movies and genres.

Isidora25 commented on src/genre/favorite-genre.service.ts:
```diff
@@ -0,0 +1,39 @@
+      where: { user: { id: userId }, genreId },
+    });
+    if (existingFavoriteGenre) {
+      await this.genreRepository.delete({ genreId });
```
если у тебя есть несколько юзеров, у которых один и тот же жанр в избранном, то нет гарантии, что удалится действительно у того юзера, который делал запрос, так что userId тут тоже нужен

Isidora25 commented on src/genre/favorite-genre.service.ts:
```diff
@@ -0,0 +1,39 @@
+    const existingFavoriteGenre = await this.genreRepository.findOne({
+      where: { user: { id: userId }, genreId },
+    });
+    if (existingFavoriteGenre) {
+      await this.genreRepository.delete({ genreId });
+      return { success: true };
+    } else {
+      const newFavoriteGenre = this.genreRepository.create({
+        user: { id: userId },
+        genreId,
+      });
+      await this.genreRepository.save(newFavoriteGenre);
+      return { success: true };
+    }
```
```suggestion
    if (existingFavoriteGenre) {
      await this.genreRepository.delete({ genreId });
   
    } else {
      const newFavoriteGenre = this.genreRepository.create({
        user: { id: userId },
        genreId,
      });
      await this.genreRepository.save(newFavoriteGenre);
      
    }
    return { success: true };
```

Isidora25 commented on src/movie/favorite-movie.service.ts:
```diff
@@ -0,0 +1,55 @@
+      where: { user: { id: userId }, movieId },
+    });
+    if (existingFavoriteMovie) {
+      await this.movieRepository.delete({ movieId });
```
те же замечания, что для жанров

Isidora25 commented on src/movie/favorite-movie.service.ts:
```diff
@@ -0,0 +1,55 @@
+    const existingFavoriteMovie = await this.movieRepository.findOne({
+      where: { user: { id: userId }, movieId },
+    });
+    if (existingFavoriteMovie.isWatched) {
+      existingFavoriteMovie.isWatched = false;
+    } else {
+      existingFavoriteMovie.isWatched = true;
+    }
```
```suggestion
const isWatched = existingFavoriteMovie.isWatched
   
      existingFavoriteMovie.isWatched = !isWatched;
    
```

Danila-Peshev committed:
Commit:
Correction for deletion by user id in genre and movie services;
Shortened the code when switching between watching movies in the service.

reshetov commented on src/genre/favorite-genre.service.ts:
```diff
@@ -0,0 +1,38 @@
+        user: { id: userId },
+        genreId,
+      });
+      await this.genreRepository.save(newFavoriteGenre);
```
А точно save отдельно нужен, сreate не сохраняет разве сразу?

reshetov commented on src/movie/favorite-movie.service.ts:
```diff
@@ -0,0 +1,51 @@
+        movieId,
+        isWatched: false,
+      });
+      await this.movieRepository.save(newFavoriteMovie);
```
см. выше
